// @react-navigation/drawer/lib/module/views/DrawerItemList.js
import * as React from 'react';
import { CommonActions, DrawerActions, useLinkBuilder } from '@react-navigation/native';
import DrawerItem from './DrawerItem';

/**
 * Component that renders the navigation list in the drawer.
 */
export default function DrawerItemList({
  state,
  navigation,
  descriptors
}) {
  const buildLink = useLinkBuilder();
  return state.routes.map((route, i) => {
    const focused = i === state.index;
    const {
      title,
      drawerLabel,
      drawerIcon,
      drawerActiveTintColor,
      drawerInactiveTintColor,
      drawerActiveBackgroundColor,
      drawerInactiveBackgroundColor,
      drawerLabelStyle,
      drawerItemStyle
    } = descriptors[route.key].options;
    return /*#__PURE__*/React.createElement(DrawerItem, {
      key: route.key,
      label: drawerLabel !== undefined ? drawerLabel : title !== undefined ? title : route.name,
      icon: drawerIcon,
      focused: focused,
      activeTintColor: drawerActiveTintColor,
      inactiveTintColor: drawerInactiveTintColor,
      activeBackgroundColor: drawerActiveBackgroundColor,
      inactiveBackgroundColor: drawerInactiveBackgroundColor,
      labelStyle: drawerLabelStyle,
      style: drawerItemStyle,
      to: buildLink(route.name, route.params),
      onPress: () => {
        navigation.dispatch({ ...(focused ? DrawerActions.closeDrawer() : CommonActions.navigate(route.name)),
          target: state.key
        });
      }
    });
  });
}
